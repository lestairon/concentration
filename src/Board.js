import React, { useEffect } from "react";
import Card from "./Card";
import Counter from "./Counter";
import Menu from "./Menu";
import { useSelector, useDispatch } from "react-redux";
import {
  createBoard,
  toggleDisabled,
  setSolved,
  incrementMoves,
  toggleMenu,
  resetFlipped
} from "./actions";

const Board = ({ numberOfCards }) => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(createBoard(numberOfCards));
  }, []);

  const cards = useSelector(({ state }) => state.cards);
  const flippedCards = useSelector(({ state }) => state.flippedCards);
  const disabled = useSelector(({ state }) => state.disabled);
  const solved = useSelector(({ state }) => state.solved);
  const moveCount = useSelector(({ state }) => state.moveCount);
  const showMenu = useSelector(({ state }) => state.showMenu);

  const cleanState = () => {
    dispatch(createBoard(numberOfCards));
  };

  const validateFlipped = () => {
    const [card1, card2] = flippedCards
      .slice(-2)
      .map(card => cards.find(x => x.key === card));
    if (card1 && card2 && card1.key !== card2.key) {
      dispatch(toggleDisabled());
      if (card1.value === card2.value) {
        dispatch(setSolved([card1.key, card2.key]));
      }
      dispatch(incrementMoves());
      setTimeout(() => {
        dispatch(resetFlipped());
        dispatch(toggleDisabled());
      }, 500);
    }
  };

  useEffect(validateFlipped, [flippedCards]);

  useEffect(() => {
    if (solved.length === cards.length) {
      dispatch(toggleMenu());
    }
  }, [solved]);

  return (
    <div className="board">
      <Counter moveCount={moveCount} />
      {showMenu ? (
        <Menu cleanState={cleanState} />
      ) : (
        cards.map(({ value, key }) => (
          <Card number={value} key={key} id={key} disabled={disabled} />
        ))
      )}
    </div>
  );
};
export default Board;
