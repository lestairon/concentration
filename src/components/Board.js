import React, { useEffect } from "react";
import Card from "./Card";
import Counter from "./Counter";
import Menu from "./Menu";
import PropTypes from "prop-types";
import { useSelector, useDispatch } from "react-redux";
import {
  createBoard,
  toggleDisabled,
  setSolved,
  incrementMoves,
  toggleMenu,
  resetFlipped
} from "../actions";

const Board = ({ numberOfCards, order }) => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(createBoard({ numberOfCards, order }));
  }, []);

  const cards = useSelector(({ state }) => state.cards);
  const flippedCards = useSelector(({ state }) => state.flippedCards);
  const solved = useSelector(({ state }) => state.solved);
  const moveCount = useSelector(({ state }) => state.moveCount);
  const showMenu = useSelector(({ state }) => state.showMenu);

  const cleanState = () => {
    dispatch(createBoard({ numberOfCards }));
  };

  const validateFlipped = () => {
    const [card1, card2] = flippedCards
      .slice(-2)
      .map(card => cards.find(object => object.key === card));
    if (card1 && card2 && card1.key !== card2.key) {
      dispatch(toggleDisabled());
      if (card1.value === card2.value) {
        dispatch(setSolved([card1.key, card2.key]));
      }
      dispatch(incrementMoves());
      setTimeout(() => {
        dispatch(resetFlipped());
        dispatch(toggleDisabled());
      }, 500);
    }
  };

  useEffect(validateFlipped, [flippedCards]);

  useEffect(() => {
    if (solved.length === cards.length) {
      dispatch(toggleMenu());
    }
  }, [solved]);

  return (
    <div className="board">
      <Counter moveCount={moveCount} />
      <div className="container">
        {showMenu ? (
          <Menu cleanState={cleanState} />
        ) : (
          cards.map(({ value, key }) => (
            <Card number={value} key={key} id={key} />
          ))
        )}
      </div>
    </div>
  );
};

Board.propTypes = {
  numberOfCards: PropTypes.number.isRequired,
  order: PropTypes.bool
};

export default Board;
