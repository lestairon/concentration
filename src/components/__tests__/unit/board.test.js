import Board from "../../Board";
import React from "react";
import { render, cleanup } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import rootReducer from "../../../reducers";

afterEach(cleanup);

const mockStore = createStore(rootReducer);

const component = numberofCards =>
  render(
    <Provider store={mockStore}>
      <Board numberOfCards={numberofCards} />
    </Provider>
  );

describe("correct functionality of the board", () => {
  it("displays the board correctly", () => {
    const { container } = component(3);
    expect(container).toMatchSnapshot();
  });
  it("shows no cards if the number is 0", () => {
    const { container } = component(0);
    expect(container).toMatchSnapshot();
  });
});
