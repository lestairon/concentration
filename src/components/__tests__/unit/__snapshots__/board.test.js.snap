// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correct functionality of the board displays the board correctly 1`] = `
<div>
  <div
    class="board"
  >
    <h3
      class="counter"
    >
      0
    </h3>
    <div
      class="container"
    >
      <button
        class="card"
        style="box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);"
      />
      <button
        class="card"
        style="box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);"
      />
      <button
        class="card"
        style="box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);"
      />
      <button
        class="card"
        style="box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);"
      />
      <button
        class="card"
        style="box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);"
      />
      <button
        class="card"
        style="box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);"
      />
    </div>
  </div>
</div>
`;

exports[`correct functionality of the board shows no cards if the number is 0 1`] = `
<div>
  <div
    class="board"
  >
    <h3
      class="counter"
    >
      0
    </h3>
    <div
      class="container"
    >
      <button>
        Restart
      </button>
    </div>
  </div>
</div>
`;
